[Murad Durrani] ip

muraddurrani opened the pull request:
# ACE
> "Remove the emotional and physical clutter from your life so you can soar." - [Judith Orloff](https://www.fillingthejars.com/quotes-about-decluttering-the-mind/)

Ace removes clutter from your life by remembering the things you need to do so you don't have to. Ace is:
- Text-based
- Lightweight
- Completely **FREE!**

All you need to do to get started is:
1. Download it.
2. Launch the program.
3. Add tasks. For example, you tell Ace you need to finish your homework with the command `todo homework`.
4. Let Ace do the rest! :sunglasses:

Features:
- [x] Managing to-dos, events and deadlines
- [x] Finding tasks by keyword
- [ ] Sorting tasks by dates (soon to come)

Feel free to look through the code to see how Ace works. Here's a snippet:
```java
public void run() {
    boolean exit = false;
    ui.displayWelcome();

    while (!exit) {
        try {
            String strCommand = ui.readCommand();
            Command command = Parser.parse(strCommand);
            command.execute(taskList, ui, storage);
            exit = command.isExit();
        } catch (DukeException e) {
            ui.displayError(e.getMessage());
        } finally {
            ui.displayLine();
        }
    }
}
```

Commit:
Add Gradle support

muraddurrani committed:
Implement increment Level-1

muraddurrani committed:
Implement increment Level-2.

muraddurrani committed:
Implement increment Level-3

muraddurrani committed:
Update to Increment Level-3.

muraddurrani committed:
Implement Level-4

muraddurrani committed:
Added javadoc comments to classes and methods.

muraddurrani committed:
Implement Level-5

muraddurrani committed:
Added new Exception classes as subclasses of DukeException.

muraddurrani committed:
Implement Level-6

muraddurrani committed:
Update implementation of existing functionality to be more consistent with OOP principles

muraddurrani committed:
Implement increment Level-7

muraddurrani committed:
Implement increment Level-8

muraddurrani committed:
Merge branch 'branch-Level-8'

muraddurrani committed:
Fix bug changes that cause Task toString method to format incorrectly

muraddurrani committed:
Implement increment A-Packages

muraddurrani committed:
Implement increment A-JUnit

muraddurrani committed:
Implement increment A-Jar

muraddurrani committed:
Implement increment A-JavaDoc

muraddurrani committed:
Implement increment A-CodingStandard

muraddurrani committed:
Update styling to be consistent with coding standard.

muraddurrani committed:
Implement increment Level-9

muraddurrani committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/InvalidArgumentException.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/ui/Ui.java

muraddurrani committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/ui/Ui.java

muraddurrani committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

muraddurrani committed:
Implement increment A-CheckStyle

muraddurrani committed:
Implement increment Level-10

B1LLP4RK commented on data/tasks.txt:
```diff
@@ -0,0 +1,4 @@
+[D][ ] task (by: 10/2/22 03:30)
+[D][X] task (by: 11/3/22 04:20)
+[T][ ] party
+[E][X] shower (at: 12/12/20 03:45)
```
Is there a reason you did not include this file in .gitignore?

B1LLP4RK commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @param input The user's input.
+     * @return Duke's response.
+     */
```
I like your clear and concise javadoc. Great job! 👍

B1LLP4RK commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,147 @@
+        String keyword = splitCommand[0];
+        Command command;
+
+        switch (keyword) {
```
I like the idea of using switch case statement here. :D

B1LLP4RK commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,30 @@
+     * @param storage The Storage object responsible for saving the change.
+     * @throws DukeException If an error is encountered during execution.
+     */
+    public abstract String execute(TaskList taskList, Ui ui, Storage storage) throws DukeException;
```
Great idea to throw error at the `Command` class's `execute` method! 💯

B1LLP4RK reviewed (commented):
Succinct code with apt names for variables and methods. Was a pleasure to read the code :)
LGTM!

sheimoria commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @param index The index of the task to be deleted.
+     */
+    public DeleteCommand(int index) {
+        this.index = index;
+    }
```
Could consider using a scratch variable `i` to indicate its short scope and perhaps reduce the use of the `this` keyword?
```suggestion
    public DeleteCommand(int i) {
        index = i;
    }
```

sheimoria commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,147 @@
+        String keyword = splitCommand[0];
+        Command command;
+
+        switch (keyword) {
+        case "bye":
+            command = new ExitCommand();
```
Good job configuring the `switch` statement's indentation to fit the coding standard!

sheimoria commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @param filepath The filepath in which the task list is stored and retrieved from.
+     */
+    public Duke(String filepath) {
```
If I'm not wrong, `filepath` should be in camelCase (as shown in the example code in Week 3's task `A-MoreOOP`)
```suggestion
    public Duke(String filePath) {
```

sheimoria commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,45 @@
+
+public class ParserTest {
+    @Test
+    public void parse_ValidInputs_CreateCommands() {
```
According to the [coding standard](https://se-education.org/guides/conventions/java/intermediate.html), test methods should also be in camelCase
```suggestion
    public void parse_validInputs_createCommands() {
```

sheimoria reviewed (commented):
Just some cosmetic nitpicking, but otherwise very well done! LGTM 👍

muraddurrani committed:
Implement increment A-Assertions

muraddurrani committed:
Increment inplement A-CodeQuality

Storage and Parser contain a few very long methods with several
levels of abstraction and nested blocks.

Decomposing these methods into smaller methods with smaller
functionality would make the code more readable.

Let's reduce nested loops in methods like parse() and load()
by creating new methods to handle those functionalities.

muraddurrani committed:
Merge pull request #1 from muraddurrani/branch-A-CodeQuality

Increment inplement A-CodeQuality

muraddurrani committed:
Merge branch 'master' into branch-A-Assertions

muraddurrani committed:
Merge pull request #2 from muraddurrani/branch-A-Assertions

Implement increment A-Assertions

muraddurrani committed:
Implement Increment C-DetectDuplicate

Users can add multiple tasks with the same name.

Users may accidentally add duplicate tasks without noticing, which
can cause confusion.

Let's detect when a duplicate task is added and give a warning
to the user.

Checking for duplicates is performed before the task is added so
the whole task list does not need to be iterated over to find
duplicates.

muraddurrani committed:
Revert "Implement Increment C-DetectDuplicate"

This reverts commit bfff6e0e1576d6876efc23eef91a03f18730afa1.

muraddurrani committed:
Implement Increment C-DetectDuplicate

Users can add multiple tasks with the same name.

Users may accidentally add duplicate tasks without noticing, which
can cause confusion.

Let's detect when a duplicate task is added and give a warning
to the user.

Checking for duplicates is performed before the task is added so
the whole task list does not need to be iterated over to find
duplicates.

muraddurrani committed:
Update parsing error messages to be more descriptive

muraddurrani committed:
Implement increments A-BetterGui and A-UserGuide

muraddurrani committed:
Update ReadMe.md
